src/main/java/org/candlepin/subscriptions/marketplace/MarketplaceSubscriptionIdProvider.java

```diff
@@ -0,0 +1,77 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Red Hat trademarks are not licensed under GPLv3. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.subscriptions.marketplace;
+
+import org.candlepin.subscriptions.db.SubscriptionRepository;
+import org.candlepin.subscriptions.db.model.ServiceLevel;
+import org.candlepin.subscriptions.db.model.Subscription;
+import org.candlepin.subscriptions.db.model.Usage;
+import org.candlepin.subscriptions.tally.UsageCalculation;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import java.util.Optional;
+
+/**
+ * Class responsible for searching Swatch database for subscriptionIds corresponding to usage keys and if
+ * none is found, delegates fetching the subscriptionId to the {@link MarketplaceSubscriptionIdCollector}.
+ */
+@Component
+public class MarketplaceSubscriptionIdProvider {
+
+    private final MarketplaceSubscriptionIdCollector collector;
+    private final SubscriptionRepository subscriptionRepo;
+
+    @Autowired
+    public MarketplaceSubscriptionIdProvider(MarketplaceSubscriptionIdCollector collector,
+        SubscriptionRepository subscriptionRepo) {
+        this.collector = collector;
+        this.subscriptionRepo = subscriptionRepo;
+    }
+
+    public Optional<String> findSubscriptionId(String accountNumber, UsageCalculation.Key usageKey) {
+        Assert.isTrue(Usage._ANY != usageKey.getUsage(), "Usage cannot be _ANY");
+        Assert.isTrue(ServiceLevel._ANY != usageKey.getSla(), "Service Level cannot be _ANY");
+
+        Optional<Subscription> result =
+            subscriptionRepo.findSubscriptionByAccountAndUsageKey(accountNumber, usageKey);
+
+        if (result.isPresent()) {
+            Subscription s = result.get();
+
+            /* If we have the subscription, but the marketplace subscription ID didn't come with the initial
+               fetch from SubscriptionService, then we call out to the MarketplaceSubscriptionIdCollector
+               to fetch the ID directly from Marketplace.  Then we'll update the Subscription record so we
+               only need to do the fetch once. */
+            if (!StringUtils.hasText(s.getMarketplaceSubscriptionId())) {
+                String missingId = collector.fetchSubscriptionId(accountNumber, usageKey);
+                s.setMarketplaceSubscriptionId(missingId);
+                subscriptionRepo.save(s);
+            }
+            return Optional.of(s.getMarketplaceSubscriptionId());
+        }
+
+        return Optional.empty();
+    }
```

Comment:
I think this logic needs to be reworked slightly. Instead of applying missing marketplace subscription ID information to an existing subscription record, it should fetch a whole subscription record from the marketplace subscription ID collector (I guess we should rename to MarketplaceSubscriptionCollector). 

If any matches already exist in our DB:
 - pick the first one.
 - log a warning if more than one exists.

else:

 - find matches via the MarketplaceSubscriptionCollector
 - save them to the DB
 - pick the first
 - log a warning if more than one match exists

It also occurs to me now that this method should accept a start date and end date and should filter out subscriptions that aren't valid for the passed range.

Reply:
Okay, I've updated this, but it does change the semantics of the code some (discussed on IRC but recording here for posterity).  Originally, the code assumed there are Subscriptions that might be missing marketplace subscription IDs and updated those subscriptions with missing IDs by fetching via MarketplaceSubscriptionIdCollector.  Now, the code looks for subscriptions and if it finds them, **it assumes the marketplace ID is present**.  If no matches are present, subscriptions are fetched via MarketplaceSubscriptionCollector (which ultimately is going to make a call to the Marketplace API), the results are persisted into the subscriptions table, and then we query the database again.